bolt-contracts/src/contracts/BoltChallenger.sol

```diff
@@ -98,92 +98,130 @@ contract BoltChallenger is IBoltChallenger {
             // We cannot open challenges for slots that are not finalized by Ethereum consensus yet.
             // This is admittedly a bit strict, since 32-slot deep reorgs are very unlikely.
             revert BlockIsNotFinalized();
         }
 
-        // Reconstruct the commitment digest: `keccak( keccak(signed tx) || le_bytes(slot) )`
-        bytes32 commitmentID = _computeCommitmentID(commitment);
+        // Check that all commitments are for the same slot and signed by the same sender
+        // and store the parsed transaction data for each commitment
+        TransactionData[] memory transactionsData = new TransactionData[](commitments.length);
+        (address txSender, address commitmentSigner, TransactionData memory firstTransactionData) =
+            _recoverCommitmentData(commitments[0]);
+
+        transactionsData[0] = firstTransactionData;
+
+        for (uint256 i = 1; i < commitments.length; i++) {
+            (address otherTxSender, address otherCommitmentSigner, TransactionData memory otherTransactionData) =
+                _recoverCommitmentData(commitments[i]);
+
+            transactionsData[i] = otherTransactionData;
+
+            // check that all commitments are for the same slot
+            if (commitments[i].slot != targetSlot) {
+                revert UnexpectedMixedSlots();
+            }
+
+            // check that all commitments are signed by the same sender
+            if (otherTxSender != txSender) {
+                revert UnexpectedMixedSenders();
+            }
+
+            // check that all commitments are signed by the same signer (aka "operator")
+            if (otherCommitmentSigner != commitmentSigner) {
+                revert UnexpectedMixedSigners();
+            }
+
+            // check that the nonces are strictly sequentially increasing in the bundle
+            if (otherTransactionData.nonce != transactionsData[i - 1].nonce + 1) {
+                revert UnexpectedNonceOrder();
+            }
+        }
 
-        // Verify the commitment signature against the digest
-        address commitmentSigner = ECDSA.recover(commitmentID, commitment.signature);
+        // Build the challenge ID: `keccak( keccak(signed tx 1) || keccak(signed tx 2) || ... || le_bytes(slot) )`
+        bytes32 challengeID = _computeChallengeID(commitments);
 
-        // Check that a challenge for this commitment does not already exist
-        if (challengeIDs.contains(commitmentID)) {
+        // Check that a challenge for this commitment bundle does not already exist
+        if (challengeIDs.contains(challengeID)) {
             revert ChallengeAlreadyExists();
         }
 
         // Add the challenge to the set of challenges
-        challengeIDs.add(commitmentID);
-        challenges[commitmentID] = Challenge({
-            id: commitmentID,
+        challengeIDs.add(challengeID);
+        challenges[challengeID] = Challenge({
```

Comment:
Do we clear expired challenges somewhere?

Reply:
No, they live in the storage forever, as we rely on this as an oracle for understanding if a slashing request is legit.

Reply:
Just wondering: after a challenge has been resolved can we keep its ID in the `challengeIDs` storage but remove it from `challenges`? We don't need that data anymore and can also for some gas refunds if I understood correctly (https://www.evm.codes/?fork=cancun#55).

With this, then we can perform this check on `_resolve`:
```solidity
        if (!challengeIDs.contains(challengeID)) {
            revert ChallengeDoesNotExist();
        }

		// This means the challenge has been resolved and deleted from storage successfully.
        if ( challenges[challengeID].id == bytes32(0)) {
            revert ChallengeAlreadyResolved();
        }
```

Reply:
Let me make a test to see how much gas this would save